File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> out = new ArrayList<>();
        List<Integer> currentSet = new ArrayList<>();

        generateSubsets(nums, 0, out, currentSet);

        return out;
    }

    void generateSubsets(int[] nums,
                        int i,
                        List<List<Integer>> out,
                        List<Integer> currentSet) {
        // base
        if (i == nums.length) {
            out.add(currentSet.stream().toList());
            return;
        }

        // dont include
        generateSubsets(nums, i+1, out, currentSet);

        // include num
        List<Integer> cpy = currentSet.stream().collect(Collectors.toList());
        cpy.add(nums[i]);
        generateSubsets(nums, i+1, out, cpy);
    }
}

File: Data Structures & Algorithms/remove-element/submission-1.go
func removeElement(nums []int, val int) int {
    // l, r = 0
    // for r < len(nums)
        // if nums[r] != val
            // nums[l] = nums[r]
            // l += 1
        // r += 1

    // return l

    l, r := 0, 0
    for r < len(nums) {
        if nums[r] != val {
            nums[l] = nums[r]
            l += 1
        }
        r += 1
    }

    return l
}
